packages/contracts/src/EmailSigner.sol

```diff
@@ -0,0 +1,157 @@
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice Initialize the contract with an initial owner, account salt, DKIM registry address, and verifier address.
+    /// @param _accountSalt The account salt to derive CREATE2 address of this contract.
+    /// @param _dkimRegistryAddr The address of the DKIM registry contract.
+    /// @param _verifierAddr The address of the verifier contract.
+    /// @param _templateId The templateId of the sign hash command.
+    function initialize(
+        bytes32 _accountSalt,
+        address _dkimRegistryAddr,
+        address _verifierAddr,
+        uint256 _templateId
+    ) public initializer {
+        accountSalt = _accountSalt;
+        templateId = _templateId;
+
+        if (_dkimRegistryAddr == address(0))
+            revert InvalidDKIMRegistryAddress();
+        if (_verifierAddr == address(0)) revert InvalidVerifierAddress();
+
+        dkimRegistryAddr = _dkimRegistryAddr;
+        verifierAddr = _verifierAddr;
+
+        emit DKIMRegistryUpdated(_dkimRegistryAddr);
+        emit VerifierUpdated(_verifierAddr);
+    }
+
+    /// @notice Authenticate the email sender and authorize the message in the email command.
+    /// @dev This function only verifies the authenticity of the email and command, without
+    /// handling replay protection. The calling contract should implement its own mechanisms
+    /// to prevent replay attacks, similar to how nonces are used with ECDSA signatures.
+    /// @param emailAuthMsg The email auth message containing all necessary information for authentication.
+    function verifyEmail(EmailAuthMsg memory emailAuthMsg) public view {
+        if (templateId != emailAuthMsg.templateId) revert InvalidTemplateId();
+        string[] memory signHashTemplate = new string[](2);
+        signHashTemplate[0] = "signHash";
+        signHashTemplate[1] = "{uint}";
+
+        if (
+            !IDKIMRegistry(dkimRegistryAddr).isDKIMPublicKeyHashValid(
+                emailAuthMsg.proof.domainName,
+                emailAuthMsg.proof.publicKeyHash
+            )
+        ) revert InvalidDKIMPublicKeyHash();
+
+        if (accountSalt != emailAuthMsg.proof.accountSalt)
+            revert InvalidAccountSalt();
+
+        if (
+            bytes(emailAuthMsg.proof.maskedCommand).length >
+            IVerifier(verifierAddr).commandBytes()
+        ) revert InvalidMaskedCommandLength();
+
+        if (
+            emailAuthMsg.skippedCommandPrefix >=
+            IVerifier(verifierAddr).commandBytes()
+        ) revert InvalidSkippedCommandPrefixSize();
+
+        // Construct an expectedCommand from template and the values of emailAuthMsg.commandParams.
+        string memory trimmedMaskedCommand = CommandUtils.removePrefix(
+            emailAuthMsg.proof.maskedCommand,
+            emailAuthMsg.skippedCommandPrefix
+        );
+        string memory expectedCommand = "";
+        for (uint stringCase = 0; stringCase < 3; stringCase++) {
```

Comment:
Note to ask Sora unless you know, I do not understand why we have `stringCase < 3` and we also revert on `if (stringCase == 2)` - doesn’t impact this PR